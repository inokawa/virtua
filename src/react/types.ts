import { CSSProperties, ReactNode } from "react";

export type ViewportComponentAttributes = Pick<
  React.HTMLAttributes<HTMLElement>,
  "className" | "style" | "id" | "role" | "tabIndex" | "onKeyDown" | "onWheel"
> &
  React.AriaAttributes;

export interface CustomContainerComponentProps {
  style: CSSProperties;
  children: ReactNode;
}

export type CustomContainerComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomContainerComponentProps> &
    React.RefAttributes<any>
>;

/**
 * Props of customized item component for {@link Virtualizer} or {@link WindowVirtualizer}.
 */
export interface CustomItemComponentProps {
  style: CSSProperties;
  index: number;
  children: ReactNode;
}

export type CustomItemComponent = React.ForwardRefExoticComponent<
  React.PropsWithoutRef<CustomItemComponentProps> & React.RefAttributes<any>
>;
