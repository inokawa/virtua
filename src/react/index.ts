export { VList } from "./VList.js";
export type { VListProps, VListHandle } from "./VList.js";
export { Virtualizer } from "./Virtualizer.js";
export type { VirtualizerProps, VirtualizerHandle } from "./Virtualizer.js";
export { WindowVirtualizer } from "./WindowVirtualizer.js";
export type {
  WindowVirtualizerProps,
  WindowVirtualizerHandle,
} from "./WindowVirtualizer.js";
export { VGrid as experimental_VGrid } from "./VGrid.js";
export type {
  VGridItemResize,
  VGridProps,
  VGridHandle,
  CustomCellComponent,
  CustomCellComponentProps,
} from "./VGrid.js";
export type * from "./types.js";
