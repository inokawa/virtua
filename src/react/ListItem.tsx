import {
  memo,
  useRef,
  useMemo,
  CSSProperties,
  ReactElement,
  ReactNode,
} from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { ItemResizeObserver } from "../core/resizer";
import { refKey } from "./utils";
import { isRTLDocument } from "../core/environment";
import { CustomItemComponent } from "./types";
import { NULL } from "../core/utils";

interface ListItemProps {
  _children: ReactNode;
  _resizer: ItemResizeObserver;
  _index: number;
  _offset: number;
  _hide: boolean;
  _as: "div" | CustomItemComponent;
  _isHorizontal: boolean;
  _isSSR: boolean | undefined;
}

/**
 * @internal
 */
export const ListItem = memo(
  ({
    _children: children,
    _resizer: resizer,
    _index: index,
    _offset: offset,
    _hide: hide,
    _as: Element,
    _isHorizontal: isHorizontal,
    _isSSR: isSSR,
  }: ListItemProps): ReactElement => {
    const ref = useRef<HTMLDivElement>(NULL);

    // The index may be changed if elements are inserted to or removed from the start of props.children
    useIsomorphicLayoutEffect(() => resizer(ref[refKey]!, index), [index]);

    const style = useMemo((): CSSProperties => {
      const style: CSSProperties = {
        position: hide && isSSR ? undefined : "absolute",
        [isHorizontal ? "height" : "width"]: "100%",
        [isHorizontal ? "top" : "left"]: 0,
        [isHorizontal ? (isRTLDocument() ? "right" : "left") : "top"]: offset,
        visibility: !hide || isSSR ? "visible" : "hidden",
      };
      if (isHorizontal) {
        style.display = "flex";
      }
      return style;
    }, [offset, hide, isSSR, isHorizontal]);

    if (typeof Element === "string") {
      return (
        <Element ref={ref} style={style}>
          {children}
        </Element>
      );
    } else {
      return (
        <Element ref={ref} style={style} index={index}>
          {children}
        </Element>
      );
    }
  }
);
