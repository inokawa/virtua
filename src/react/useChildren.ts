import { ReactElement, ReactNode, useMemo } from "react";
import { ItemElement, flattenChildren } from "./utils.js";

/**
 * @internal
 */
export const useChildren = <T>(
  children: ReactNode | ((data: T, i: number) => ReactElement),
  data: ArrayLike<T> | undefined
) => {
  return useMemo((): [(i: number) => ItemElement, number] => {
    if (typeof children === "function") {
      return [(i) => children(data![i]!, i), data!.length];
    }
    // Memoize element array
    const _elements = flattenChildren(children);
    return [(i) => _elements[i]!, _elements.length];
  }, [children, data]);
};
