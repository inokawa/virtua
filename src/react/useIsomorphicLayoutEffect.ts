import { useEffect, useLayoutEffect } from "react";
import { isBrowser } from "../core";

/**
 * https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
 * @internal
 */
export const useIsomorphicLayoutEffect = isBrowser
  ? useLayoutEffect
  : useEffect;
