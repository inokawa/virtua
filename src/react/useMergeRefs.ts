import { Ref, MutableRefObject, useCallback } from "react";

/**
 * @internal
 * Merge multiple refs into a single ref callback.
 */
export function useMergeRefs<T>(...refs: Array<Ref<T> | undefined>): Ref<T> {
  return useCallback((instance: T | null) => {
    for (const ref of refs) {
      ref && updateRef(ref, instance);
    }
  }, refs);
}

function updateRef<T>(ref: Ref<T>, instance: T | null): void {
  if (typeof ref === "function") {
    ref(instance);
  } else if (ref) {
    (ref as MutableRefObject<T | null>).current = instance;
  }
}
