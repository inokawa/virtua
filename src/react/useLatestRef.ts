import { useRef } from "react";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect.js";
import { refKey } from "./utils.js";

/**
 * @internal
 */
export const useLatestRef = <T>(value: T) => {
  const ref = useRef<T>(value);

  useIsomorphicLayoutEffect(() => {
    ref[refKey] = value;
  }, [value]);

  return ref;
};
