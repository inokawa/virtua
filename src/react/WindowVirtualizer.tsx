import {
  ReactElement,
  ReactNode,
  useEffect,
  forwardRef,
  useImperativeHandle,
  useRef,
} from "react";
import {
  ACTION_ITEMS_LENGTH_CHANGE,
  UPDATE_VIRTUAL_STATE,
  getOverscanedRange,
  createVirtualStore,
  SCROLL_IDLE,
  UPDATE_SCROLL_END_EVENT,
} from "../core/store";
import { useIsomorphicLayoutEffect } from "./useIsomorphicLayoutEffect";
import { createWindowScroller } from "../core/scroller";
import { getKey, refKey } from "./utils";
import { useStatic } from "./useStatic";
import { useLatestRef } from "./useLatestRef";
import { createWindowResizer } from "../core/resizer";
import { CacheSnapshot } from "../core/types";
import { CustomContainerComponent, CustomItemComponent } from "./types";
import { ListItem } from "./ListItem";
import { flushSync } from "react-dom";
import { useRerender } from "./useRerender";
import { useChildren } from "./useChildren";

/**
 * Methods of {@link WindowVirtualizer}.
 */
export interface WindowVirtualizerHandle {
  /**
   * Get current {@link CacheSnapshot}.
   */
  readonly cache: CacheSnapshot;
}

/**
 * Props of {@link WindowVirtualizer}.
 */
export interface WindowVirtualizerProps {
  /**
   * Elements rendered by this component.
   *
   * You can also pass a function and set {@link WindowVirtualizerProps.count} to create elements lazily.
   */
  children: ReactNode | ((index: number) => ReactElement);
  /**
   * If you set a function to {@link WindowVirtualizerProps.children}, you have to set total number of items to this prop.
   */
  count?: number;
  /**
   * Number of items to render above/below the visible bounds of the list. Lower value will give better performance but you can increase to avoid showing blank items in fast scrolling.
   * @defaultValue 4
   */
  overscan?: number;
  /**
   * Item size hint for unmeasured items. It will help to reduce scroll jump when items are measured if used properly.
   *
   * - If not set, initial item sizes will be automatically estimated from measured sizes. This is recommended for most cases.
   * - If set, you can opt out estimation and use the value as initial item size.
   */
  itemSize?: number;
  /**
   * While true is set, scroll position will be maintained from the end not usual start when items are added to/removed from start. It's recommended to set false if you add to/remove from mid/end of the list because it can cause unexpected behavior. This prop is useful for reverse infinite scrolling.
   */
  shift?: boolean;
  /**
   * If true, rendered as a horizontally scrollable list. Otherwise rendered as a vertically scrollable list.
   */
  horizontal?: boolean;
  /**
   * You can restore cache by passing a {@link CacheSnapshot} on mount. This is useful when you want to restore scroll position after navigation. The snapshot can be obtained from {@link WindowVirtualizerHandle.cache}.
   *
   * **The length of items should be the same as when you take the snapshot, otherwise restoration may not work as expected.**
   */
  cache?: CacheSnapshot;
  /**
   * A prop for SSR. If set, the specified amount of items will be mounted in the initial rendering regardless of the container size until hydrated.
   */
  ssrCount?: number;
  /**
   * Component or element type for container element.
   * @defaultValue "div"
   */
  as?: keyof JSX.IntrinsicElements | CustomContainerComponent;
  /**
   * Component or element type for item element. This component will get {@link CustomItemComponentProps} as props.
   * @defaultValue "div"
   */
  item?: keyof JSX.IntrinsicElements | CustomItemComponent;
  /**
   * Callback invoked when scrolling stops.
   */
  onScrollEnd?: () => void;
  /**
   * Callback invoked when visible items range changes.
   */
  onRangeChange?: (
    /**
     * The start index of viewable items.
     */
    startIndex: number,
    /**
     * The end index of viewable items.
     */
    endIndex: number
  ) => void;
}

/**
 * {@link Virtualizer} controlled by the window scrolling. See {@link WindowVirtualizerProps} and {@link WindowVirtualizer}.
 */
export const WindowVirtualizer = forwardRef<
  WindowVirtualizerHandle,
  WindowVirtualizerProps
>(
  (
    {
      children,
      count: renderCountProp,
      overscan = 4,
      itemSize,
      shift,
      horizontal: horizontalProp,
      cache,
      ssrCount,
      as: Element = "div",
      item: ItemElement = "div",
      onScrollEnd: onScrollEndProp,
      onRangeChange: onRangeChangeProp,
    },
    ref
  ): ReactElement => {
    Element = Element as "div";

    const [getElement, count] = useChildren(children, renderCountProp);

    const containerRef = useRef<HTMLDivElement>(null);

    const onScrollEnd = useLatestRef(onScrollEndProp);

    const isSSR = useRef(!!ssrCount);

    const [store, resizer, scroller, isHorizontal] = useStatic(() => {
      const _isHorizontal = !!horizontalProp;
      const _store = createVirtualStore(
        count,
        itemSize,
        ssrCount,
        cache,
        !itemSize
      );

      return [
        _store,
        createWindowResizer(_store, _isHorizontal),
        createWindowScroller(_store, _isHorizontal),
        _isHorizontal,
      ];
    });
    // The elements length and cached items length are different just after element is added/removed.
    if (count !== store._getItemsLength()) {
      store._update(ACTION_ITEMS_LENGTH_CHANGE, [count, shift]);
    }

    const rerender = useRerender(store);

    const [startIndex, endIndex] = store._getRange();
    const scrollDirection = store._getScrollDirection();
    const jumpCount = store._getJumpCount();
    const totalSize = store._getTotalSize();

    const items: ReactElement[] = [];

    useIsomorphicLayoutEffect(() => {
      isSSR[refKey] = false;

      // store must be subscribed first because others may dispatch update on init depending on implementation
      const unsubscribeStore = store._subscribe(
        UPDATE_VIRTUAL_STATE,
        (sync) => {
          if (sync) {
            flushSync(rerender);
          } else {
            rerender();
          }
        }
      );
      const unsubscribeOnScrollEnd = store._subscribe(
        UPDATE_SCROLL_END_EVENT,
        () => {
          onScrollEnd[refKey] && onScrollEnd[refKey]();
        }
      );

      const el = containerRef[refKey]!;
      resizer._observeRoot(el);
      scroller._observe(el);
      return () => {
        unsubscribeStore();
        unsubscribeOnScrollEnd();
        resizer._dispose();
        scroller._dispose();
      };
    }, []);

    useIsomorphicLayoutEffect(() => {
      scroller._fixScrollJump();
    }, [jumpCount]);

    useEffect(() => {
      if (!onRangeChangeProp) return;

      onRangeChangeProp(startIndex, endIndex);
    }, [startIndex, endIndex]);

    useImperativeHandle(
      ref,
      () => {
        return {
          get cache() {
            return store._getCacheSnapshot();
          },
        };
      },
      []
    );

    for (
      let [i, j] = getOverscanedRange(
        startIndex,
        endIndex,
        overscan,
        scrollDirection,
        count
      );
      i <= j;
      i++
    ) {
      const e = getElement(i);
      items.push(
        <ListItem
          key={getKey(e, i)}
          _resizer={resizer._observeItem}
          _index={i}
          _offset={store._getItemOffset(i)}
          _hide={store._isUnmeasuredItem(i)}
          _as={ItemElement as "div"}
          _children={e}
          _isHorizontal={isHorizontal}
          _isSSR={isSSR[refKey]}
        />
      );
    }

    return (
      <Element
        ref={containerRef}
        style={{
          // contain: "content",
          flex: "none", // flex style can break layout
          position: "relative",
          visibility: "hidden", // TODO replace with other optimization methods
          width: isHorizontal ? totalSize : "100%",
          height: isHorizontal ? "100%" : totalSize,
          pointerEvents: scrollDirection !== SCROLL_IDLE ? "none" : "auto",
        }}
      >
        {items}
      </Element>
    );
  }
);
