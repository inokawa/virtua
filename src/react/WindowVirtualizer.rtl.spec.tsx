import { afterEach, it, expect, describe } from "vitest";
import { cleanup } from "@testing-library/react";
import { WindowVirtualizer } from "./WindowVirtualizer.js";
import { setupResizeJsDom } from "../../scripts/spec.js";
import { render } from "../../scripts/spec-react.js";

Object.defineProperty(CSSStyleDeclaration.prototype, "direction", {
  get() {
    return "rtl";
  },
});

setupResizeJsDom({
  itemSize: { width: 100, height: 50 },
});

afterEach(cleanup);

describe("rtl", async () => {
  it("should not work in vertical", async () => {
    const { asFragment } = await render(
      <WindowVirtualizer>
        {Array.from({ length: 100 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </WindowVirtualizer>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should work in horizontal", async () => {
    const { asFragment } = await render(
      <WindowVirtualizer horizontal>
        {Array.from({ length: 100 }).map((_, i) => (
          <div key={i}>{i}</div>
        ))}
      </WindowVirtualizer>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
