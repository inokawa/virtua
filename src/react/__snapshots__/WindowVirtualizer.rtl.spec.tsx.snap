// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rtl > should not work in vertical 1`] = `
<DocumentFragment>
  <div
    style="flex: 0 0 auto; position: relative; visibility: hidden; width: 100%; height: 5000px;"
  >
    <div
      style="position: absolute; width: 100%; left: 0px; top: 0px; visibility: visible;"
    >
      <div>
        0
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 50px; visibility: visible;"
    >
      <div>
        1
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 100px; visibility: visible;"
    >
      <div>
        2
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 150px; visibility: visible;"
    >
      <div>
        3
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 200px; visibility: visible;"
    >
      <div>
        4
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 250px; visibility: visible;"
    >
      <div>
        5
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 300px; visibility: visible;"
    >
      <div>
        6
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 350px; visibility: visible;"
    >
      <div>
        7
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 400px; visibility: visible;"
    >
      <div>
        8
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 450px; visibility: visible;"
    >
      <div>
        9
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 500px; visibility: visible;"
    >
      <div>
        10
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 550px; visibility: visible;"
    >
      <div>
        11
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 600px; visibility: visible;"
    >
      <div>
        12
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 650px; visibility: visible;"
    >
      <div>
        13
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 700px; visibility: visible;"
    >
      <div>
        14
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 750px; visibility: visible;"
    >
      <div>
        15
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 800px; visibility: visible;"
    >
      <div>
        16
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 850px; visibility: visible;"
    >
      <div>
        17
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 900px; visibility: visible;"
    >
      <div>
        18
      </div>
    </div>
    <div
      style="position: absolute; width: 100%; left: 0px; top: 950px; visibility: visible;"
    >
      <div>
        19
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`rtl > should work in horizontal 1`] = `
<DocumentFragment>
  <div
    style="flex: 0 0 auto; position: relative; visibility: hidden; width: 10000px; height: 100%;"
  >
    <div
      style="position: absolute; height: 100%; top: 0px; right: 0px; visibility: visible; display: flex;"
    >
      <div>
        0
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 100px; visibility: visible; display: flex;"
    >
      <div>
        1
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 200px; visibility: visible; display: flex;"
    >
      <div>
        2
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 300px; visibility: visible; display: flex;"
    >
      <div>
        3
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 400px; visibility: visible; display: flex;"
    >
      <div>
        4
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 500px; visibility: visible; display: flex;"
    >
      <div>
        5
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 600px; visibility: visible; display: flex;"
    >
      <div>
        6
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 700px; visibility: visible; display: flex;"
    >
      <div>
        7
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 800px; visibility: visible; display: flex;"
    >
      <div>
        8
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 900px; visibility: visible; display: flex;"
    >
      <div>
        9
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 1000px; visibility: visible; display: flex;"
    >
      <div>
        10
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 1100px; visibility: visible; display: flex;"
    >
      <div>
        11
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 1200px; visibility: visible; display: flex;"
    >
      <div>
        12
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 1300px; visibility: visible; display: flex;"
    >
      <div>
        13
      </div>
    </div>
    <div
      style="position: absolute; height: 100%; top: 0px; right: 1400px; visibility: visible; display: flex;"
    >
      <div>
        14
      </div>
    </div>
  </div>
</DocumentFragment>
`;
